src/wire-protocol.ts

```diff
@@ -47,27 +47,69 @@ export interface QueryRequestHeaders {
   traceparent?: string;
 }
 
-/**
- * A response to a query.
- * @remarks
- * The QueryResponse is type parameterized so that you can treat it as a
- * a certain type if you are using Typescript.
- */
-export interface QueryResponse<T> {
+export type QueryStats = {
+  /** The amount of Transactional Compute Ops consumed by the query. */
+  compute_ops: number;
+  /** The amount of Transactional Read Ops consumed by the query. */
+  read_ops: number;
+  /** The amount of Transactional Write Ops consumed by the query. */
+  write_ops: number;
+  /** The query run time in milliseconds. */
+  query_time_ms: number;
+  /** The amount of data read from storage, in bytes. */
+  storage_bytes_read: number;
+  /** The amount of data written to storage, in bytes. */
+  storage_bytes_written: number;
+  /** The number of times the transaction was retried due to write contention. */
+  contention_retries: number;
+};
+
+export type QueryInfo = {
+  /** The last transaction time of the query. An ISO-8601 date string. */
+  txn_time: string;
+  /** A readable summary of any warnings or logs emitted by the query. */
+  summary?: string;
+  /** The value of the x-query-tags header, if it was provided. */
+  query_tags: Record<string, string>;
+  /** Stats on query performance and cost */
+  stats: QueryStats;
+};
+
+export type QuerySuccess<T> = QueryInfo & {
   /**
    * The result of the query. The data is any valid JSON value.
    * @remarks
    * data is type parameterized so that you can treat it as a
    * certain type if you are using typescript.
    */
   data: T;
-  /** Stats on query performance and cost */
-  stats: { [key: string]: number };
-  /** The last transaction time of the query. An ISO-8601 date string. */
-  txn_time: string;
-  /** A readable summary of any warnings or logs emitted by the query. */
-  summary?: string;
-}
+  /** The query's inferred static result type. */
+  static_type?: string;
+};
+
+/**
+ * A failed query response. Integrations which only want to report a human
+ * readable version of the failure can simply print out the "summary" field.
+ */
+export type QueryFailure = QueryInfo & {
+  /**
+   * The result of the query resulting in
+   */
+  error: {
+    /** A predefined code which indicates the type of error. See XXX for a list of error codes. */
+    code: string;
+    /** description: A short, human readable description of the error */
+    message: string;
+  };
+};
+
+export const isQuerySuccess = <T>(
+  res: QuerySuccess<T> | QueryFailure
+): res is QuerySuccess<T> => "data" in res;
+
+export const isQueryFailure = (
+  res: QuerySuccess<any> | QueryFailure
+): res is QueryFailure => "error" in res;
```

Comment:
we should document these and refine them - but they aren't used heavily until the fetch refactor so putning for now.